django/db/models/fields/related_descriptors.py

```diff
@@ -912,6 +912,30 @@ def get_prefetch_queryset(self, instances, queryset=None):
                 False,
             )
 
+        def count(self):
+            # If the through's target field's foreign integrity is enforced
+            # the COUNT can be performed against the through table instead of
+            # INNER JOIN'ing the target table.
+            if self.target_field.db_constraint:
```

Comment:
The optimization should also be disabled `if not connections[db].features.supports_foreign_keys`.